### PR TITLE
add logout button && no more matches for pairing popup && upload popup

### LINKED ISSUE #84: Log out button
**Describe the task that needs to be done.**

This issue was discussed in the team meeting of the user's team on 17th of March 2022 and all team members approved this issue.
There should be a way for users to log out from the application. A possible place to put the 'Log Out' button would be the `Settings` page component or the `SideBar` component.

**Describe what a solution to your proposed task might look like (and any alternatives considered).**

Logging out the user from the client-side would consist of accessing the `AuthContext` and clearing the JWT token from the global context. This would automatically take the user to the `/login` page because they cannot access the protected routes anymore. 

**Notes**

- This issue will be worked on by the next team.

### LINKED ISSUE #129: No more matches in popup
**Describe the task that needs to be done.**
Adding a final state to the pairing popup, where if a user has looped through all possible matches, then instead of the component immediately closing, it will say "no more matches" 

**Describe how a solution to your proposed task might look like (and any alternatives considered).**

image

**Notes**

### LINKED ISSUE #130: Upload Popup
**Describe the task that needs to be done.**
Moving the uploading of a user's timetable to being added through a popup component that is triggered only if a user does not have a timetable already loaded.

This would then support the future addition of the timetable component in the background.

**Describe how a solution to your proposed task might look like (and any alternatives considered).**
Would use the current layout, but likely in a dialog component similar to the pairing popup or otherwise. 

**Notes**

### PR DESCRIPTION
**Pull request by Xiaoqi510**
## Description
1. Add a 'Log Out' button in src/components/NavBar/SideBar.jsx, so that users could be able to log out from the application. 
2. Add a ''NoBuddy' popup component in  src/components/NoBuddy/ , so that if a user has looped through all possible matches, the component will not immediately closing, it will say "no more matches".
3. Add upload popup component in src/components/UploadPopUp/, so that users can upload their timetable file from local machine. Users can also preview and delete the files in the upload dialog.

resolves #129 
resolves #84 
resolves #130 
**(To member who responsible for uploading files: #130 only create upload popup component, not including submit functions. Also when you do the submit function, please consider the file storage format and remember to change acceptedFiles number.)**
## How has this been tested?
manual testing

## Screenshots
***Add logout button***
Screen Shot 2022-03-28 at 5 15 27 PM
***Add no more matches for paring dialog***
Screen Shot 2022-03-28 at 5 14 23 PM
***Add upload timetable dialog***
Screen Shot 2022-03-30 at 11 41 12 AM

## Checklist
*(Leave blank if not applicable)*

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by christopher-alba — CHANGES_REQUESTED** [review]
image
Looking great! Just one thing to fix for the snapshot testing for the nav. I think you just have to delete the Nav.test.jsx.snap  under __snapshots__ folder and re-run the tests using ```npm test``` and it should generate a new snapshot :)

**Comment by Xiaoqi510** [comment]
> image Looking great! Just one thing to fix for the snapshot testing for the nav. I think you just have to delete the Nav.test.jsx.snap under **snapshots** folder and re-run the tests using `npm test` and it should generate a new snapshot :)

Thank you for your review, I modified it, please try **npm test** again

**Review by christopher-alba — CHANGES_REQUESTED** [review]
Heya, not sure if its just on my end but I think the library you are using is a little bugged. The styles are being removed once a file is uploaded to the dropzone for some reason. 
image
I'm not too sure what you can do about this, maybe just use a standard file upload component.

**Comment by Xiaoqi510** [comment]
> Heya, not sure if its just on my end but I think the library you are using is a little bugged. The styles are being removed once a file is uploaded to the dropzone for some reason. image I'm not too sure what you can do about this, maybe just use a standard file upload component.

thank you for you suggestion, I change it from dropzone to input. please take a look

File: package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "axios": "^0.26.1",
     "emoji-picker-react": "^3.5.1",
     "history": "^5.3.0",
+    "material-ui-dropzone": "^3.5.0",
```
Comment (christopher-alba) [inline package.json]: Please uninstall the package using ```npm uninstall material-ui-dropzone``` command

**Comment by hpar461** [comment]
Code changes made to resolve issue #129 collide with code changes made in PR #139 and it will cause merge conflicts. It needs to be resolved.

**Comment by Xiaoqi510** [comment]
> Code changes made to resolve issue #129 collide with code changes made in PR #139 and it will cause merge conflicts. It needs to be resolved.

thank you for your mention, resolved the conflicts now

File: package-lock.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
         "axios": "^0.26.1",
         "emoji-picker-react": "^3.5.1",
         "history": "^5.3.0",
+        "material-ui-dropzone": "^3.5.0",
```
Comment (christopher-alba) [inline package-lock.json]: it seems the package-lock still has the dependencies of material-ui-dropzone installed, did you run the uninstall command or just delete it from the package.json?

**Comment by Xiaoqi510** [comment]
> Heya, not sure if its just on my end but I think the library you are using is a little bugged. The styles are being removed once a file is uploaded to the dropzone for some reason. image I'm not too sure what you can do about this, maybe just use a standard file upload component.

fixed

**Comment by Xiaoqi510** [comment]
> Heya, not sure if its just on my end but I think the library you are using is a little bugged. The styles are being removed once a file is uploaded to the dropzone for some reason. image I'm not too sure what you can do about this, maybe just use a standard file upload component.

fixed

File: src/pages/Upload/UploadPage.jsx [hunk 3]
```diff
@@ -1,79 +1,74 @@
...
-  Typography,
-  Link,
-  InputAdornment,
-  FormControl,
-  InputLabel,
-  OutlinedInput,
-} from '@mui/material';
+import { Button, Dialog } from '@mui/material';
+import UploadIcon from '@mui/icons-material/Upload';
+import styles from './UploadPopUp.module.css';
```
Comment (christopher-alba) [inline src/pages/Upload/UploadPage.jsx]:
```suggestion
import styles from './UploadPage.module.css';
```
image
I seem to be getting this error in the local environment, maybe try out this suggestion please.

**Review by christopher-alba — CHANGES_REQUESTED** [review]

**Review by christopher-alba — DISMISSED** [review]
Nice work! LGTM!

**Review by yuchendong13 — DISMISSED** [review]
LGTM!

File: src/context/AuthContext.jsx [hunk 4]
```diff
@@ -21,6 +21,7 @@ export function AuthContextProvider({ children }) {
   const [tokenId, setTokenId] = useState('');
 
   // Clears the JWT, which triggers the useEffect() below to set authenticated to false.
+
```
Comment (R055A) [inline src/context/AuthContext.jsx]:
```suggestion
```
This change is not required and adds an unecessary space after a comment

**Review by R055A — CHANGES_REQUESTED** [review]
LGTM, aside for just one very small suggestion

**Review by Tvo465 — DISMISSED** [review]
Looks good. Nice work :)

**Review by qzho236 — DISMISSED** [review]

**Review by yuchendong13 — APPROVED** [review]

**Review by R055A — APPROVED** [review]

File: src/context/AuthContext.jsx (hunk 4 above)
Comment (R055A) [reply src/context/AuthContext.jsx]: Perfect. Thank you

**Comment by R055A** [comment]
LGTM!

**Review by christopher-alba — APPROVED** [review]
Nice!

**Pull request merged**